src/unix/io_uring.c

```diff
@@ -0,0 +1,1340 @@
+                    fdesc_put(old);
+                iour->files[offset + i] = f;
+            } else {
+                if (i > 0)
+                    ret = i;
+                break;
+            }
+        }
+    }
+    iour_unlock(iour);
+    if (!ret)
+        ret = count;
+    return ret;
+}
+
+static boolean iour_submit(io_uring iour, struct io_uring_sqe *sqe)
+{
+    iour_debug("opcode %d, flags 0x%x, user_data %ld", sqe->opcode, sqe->flags,
+        sqe->user_data);
+    fdesc f = 0;
+    s32 res;
+    if (sqe->flags & ~(IOSQE_FIXED_FILE | IOSQE_ASYNC)) {
+        /* non-supported flags */
+        res = -EINVAL;
+        goto complete;
+    }
+    switch(sqe->opcode) {
+    case IORING_OP_READV:
+    case IORING_OP_WRITEV:
+    case IORING_OP_READ_FIXED:
+    case IORING_OP_WRITE_FIXED:
+    case IORING_OP_POLL_ADD:
+    case IORING_OP_READ:
+    case IORING_OP_WRITE:
+        if (sqe->flags & IOSQE_FIXED_FILE) {
+            iour_lock(iour);
+            int fd = sqe->fd;
+            if ((fd >= 0) && (fd < iour->file_count)) {
+                f = iour->files[fd];
+                if (f)
+                    f->refcnt++;
+            }
+            iour_unlock(iour);
+        } else
+            f = fdesc_get(current->p, sqe->fd);
+        if (!f) {
+            res = -EBADF;
+            goto complete;
+        }
+        break;
+    default:
+        break;
+    }
+    switch (sqe->opcode) {
+    case IORING_OP_NOP:
+        res = 0;
+        goto complete;
+    case IORING_OP_READV:
+    case IORING_OP_WRITEV: {
+        if (sqe->buf_index) {
+            res = -EINVAL;
+            goto complete;
+        }
+        struct iovec *iov = pointer_from_u64(sqe->addr);
+        u32 len = sqe->len;
+        boolean write = (sqe->opcode == IORING_OP_WRITEV);
+        if (!validate_iovec(iov, len, write)) {
+            res = -EFAULT;
+            goto complete;
+        }
+        iour_iov(iour, f, write, iov, len, sqe->off, sqe->user_data);
+        break;
+    }
+    case IORING_OP_READ_FIXED:
+    case IORING_OP_WRITE_FIXED:
+        res = 0;
+        iour_lock(iour);
+        u16 buf_index = sqe->buf_index;
+        if (buf_index >= iour->buf_count) {
+            if (iour->buf_count != 0)
+                res = -EINVAL;
+            else
+                res = -EFAULT;
+        } else {
+            struct iovec *iov = &iour->bufs[buf_index];
+            void *buf = pointer_from_u64(sqe->addr);
+            u32 len = sqe->len;
+            boolean write = sqe->opcode == IORING_OP_WRITE_FIXED;
+
+            /* If we have the capability to pin memory pages, as done in Linux
+             * with pin_user_pages(), we will validate the memory at buffer
+             * registration time and avoid validating it here. */
+            if ((buf < iov->iov_base) || (u64_from_pointer(buf) + len >
+                    u64_from_pointer(iov->iov_base) + iov->iov_len) ||
+                    !validate_user_memory(buf, len, write)) {
```

Comment:
!write

Reply:
Right. Also found a mistake in `io_uring_enter()`, when validating `sig`, it should be `validate_user_memory(sig, sizeof(*sig), false)`.